src/propagate_precision.cpp

```diff
@@ -0,0 +1,191 @@
+#include <migraphx/propagate_precision.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/eliminate_convert.hpp>
+#include <unordered_set>
+#include <unordered_map>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace {
+struct precision
+{
+    shape::type_t type;
+
+    friend bool operator==(const precision& xp, const precision& yp) { return xp.type == yp.type; }
+    friend bool operator<(const precision& xp, const precision& yp)
+    {
+        bool is_less = false;
+        shape::visit(xp.type, [&](auto x) {
+            shape::visit(yp.type, [&](auto y) {
+                if(x.is_integral() != y.is_integral())
+                    return;
+                if(x.is_integral())
+                {
+                    if(x.is_unsigned() != y.is_unsigned() and x.size() == y.size())
+                        is_less = y.is_unsigned();
+                    else
+                        is_less = x.size() < y.size();
+                }
+                else
+                {
+                    is_less = x.size() < y.size();
+                }
+            });
+        });
+        return is_less;
+    }
+    friend bool operator!=(const precision& xp, const precision& yp) { return not(xp == yp); }
+    friend bool operator>(const precision& xp, const precision& yp) { return yp < xp; }
+    // This is not totally ordered
+    friend bool operator<=(const precision& xp, const precision& yp)
+    {
+        return (xp < yp) or (xp == yp);
+    }
+    friend bool operator>=(const precision& xp, const precision& yp)
+    {
+        return (xp > yp) or (xp == yp);
+    }
```

Comment:
This might seem like an odd ask but why not make these xor vs or? If one is true then the other doesn't matter as the result shouldn't be true. Anyway adding xor instead of or here can speed things up / check for errors?